wire-ios-data-model/Source/E2EIdentity/MLSConversationVerificationStatusProvider.swift

```diff
@@ -0,0 +1,97 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+// sourcery: AutoMockable
+public protocol MLSConversationVerificationStatusProviderInterface {
+
+    func updateStatus(_ groupID: MLSGroupID) async throws
+
+}
+
+public class MLSConversationVerificationStatusProvider: MLSConversationVerificationStatusProviderInterface {
+
+    // MARK: - Properties
+
+    private var e2eIVerificationStatusService: E2eIVerificationStatusServiceInterface
+    private var syncContext: NSManagedObjectContext
+
+    // MARK: - Life cycle
+
+    public init(
+        e2eIVerificationStatusService: E2eIVerificationStatusServiceInterface,
+        syncContext: NSManagedObjectContext) {
+            self.e2eIVerificationStatusService = e2eIVerificationStatusService
+            self.syncContext = syncContext
+        }
+
+    // MARK: - Public interface
+
+    public func updateStatus(_ groupID: MLSGroupID) async throws {
+        guard let conversation = await syncContext.perform({
+            ZMConversation.fetch(with: groupID, in: self.syncContext)
+        }) else {
+            throw E2eIVerificationStatusService.E2eIVerificationStatusError.missingConversation
+        }
+        do {
+            let coreCryptoStatus = try await e2eIVerificationStatusService.getConversationStatus(groupID: groupID)
+            await syncContext.perform {
+                self.updateStatusAndNotifyUserIfNeeded(newStatusFromCC: coreCryptoStatus, conversation: conversation)
+            }
+        } catch {
+            throw error
+        }
+    }
+
+    // MARK: - Helpers
+
+    private func updateStatusAndNotifyUserIfNeeded(newStatusFromCC: MLSVerificationStatus,
+                                                   conversation: ZMConversation) {
+        guard let currentStatus = conversation.mlsVerificationStatus else {
+            return conversation.mlsVerificationStatus = newStatusFromCC
+        }
+
+        let newStatus = getActualNewStatus(newStatusFromCC: newStatusFromCC, currentStatus: currentStatus)
+        guard newStatus != currentStatus else {
+            return
+        }
+        conversation.mlsVerificationStatus = newStatus
+        // TODO: check conditions - https://wearezeta.atlassian.net/browse/WPB-3233
+        if newStatus == .degraded || newStatus == .verified {
+            notifyUserAboutStateChanges(newStatus, in: conversation)
+        }
+    }
+
+    private func getActualNewStatus(newStatusFromCC: MLSVerificationStatus,
+                                    currentStatus: MLSVerificationStatus) -> MLSVerificationStatus {
+        switch (newStatusFromCC, currentStatus) {
+        case (.notVerified, .verified):
+            return .degraded
+        case(.notVerified, .degraded):
+            return .degraded
```

Comment:
```suggestion
        case (.notVerified, .verified), (.notVerified, .degraded):
            return .degraded
```

Reply:
I kind of understand why the status is `degraded` when going from `verified` to `notVerified`. But why is it `degraded` when going from `degraded` to `notVerified` ?

Reply:
I would prefer to leave them on separate lines. In my opinion, it’s more readable.
The new state comes from the CC, and there is no `degraded` state on the CC. This means that CC only knows whether the conversation is verified or not. We calculate the degraded state on the client side. If it happens that the conversation becomes degraded, we should only change the status to verified.

Reply:
I understand that we have two different statuses:
- The core crypto status -> (`.verified, .notVerified, .notEnabled`)
- The conversation status -> (`.verified, .notVerified, .degraded`)

And we choose to represent them both by the same type: `MLSVerificationStatus`. I find that a bit confusing when reading through the code, but unfortunately I can't think of a better way to do it. Any thoughts on this?